common/collection/concurrentQueue.go

```diff
@@ -0,0 +1,87 @@
+// Copyright (c) 2020 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package collection
+
+import (
+	"sync"
+)
+
+type (
```

Comment:
out of curiosity, why not use a chan or a wrapper on top of chan?

Reply:
The requirement for the implementation is that it should potentially have infinite size instead of a fixed size, otherwise the Add() method will be blocking. And if Add() is blocking, then the following situation will occur:
- Assume there is only one worker goroutine and the queue is currently full. The worker goroutine executed a task and decides to redispatch this task. It calls Add(), but since the queue is full, it got blocked. Now basically there's no goroutine for processing tasks and result in a deadlock.
- If the queue has a fixed size and tasks from one domain occupies the entire queue. This blocks tasks from other domains (with a high priority) being processed and defeat the purpose of multi-tenant task processing. 
Right now the assumption here is that host memory is large enough to hold all the tasks and we can use an infinite size queue as the re-dispatch queue. Of course this is not true and this problem will be addressed by the task queue split & merge work (basically offload some tasks under high memory pressure).

Yes there are ways around the situation I mentioned above, for example, use a new goroutine from adding tasks to the re-dispatch queue and probably having one re-dispatch queue for each task priority. But right now those are not necessary and using a list solves all the problems. This, combined with the later task queue split & merge work, should not result in any memory issue.

Reply:
if the idea is to have unlimited space 
![image](https://user-images.githubusercontent.com/8762893/77395266-71186900-6d5e-11ea-85d2-0ea7205516e5.png)

then why not using a list? (rather than a slice)

Reply:
I tried both implementations actually, and benchmark shows slice has better performance. 1083 ns/op when using slice, while 1329 ns/op when using list. Slice also incurs less allocs per op. I think that's because we are storing small object in the queue (`interface{}`) so the overhead for maintaining a linked list is too high.

Reply:
just keep in mind that using slice may result in more mem consumption.
to me 1000+ vs 1300+ ns is not a huge difference.

Reply:
Yes, I am aware of that fact that doing `items[1:]` will not free the memory for the first item. But it will eventually be garbage collected during slice reallocation, so the memory will not grow unbounded. 

I will monitor the mem usage in prod and if it turns out to be a problem, I will switch to list implementation. If there's no mem issue, then I prefer the one with better performance in terms of time.